cpp/src/rolling/window_utils.cu

```diff
@@ -0,0 +1,255 @@
+    auto const d_offset            = dynamic_cast<ScalarType const&>(offset).data();
+    if (window_type == window_type::LEFT_CLOSED) {
+      thrust::copy_n(rmm::exec_policy(stream),
+                     cudf::detail::make_counting_transform_iterator(
+                       0,
+                       distance_kernel<T, OffsetType, op_t<Which, window_type::LEFT_CLOSED>>{
+                         d_length, d_offset, input_begin, input_end}),
+                     input.size(),
+                     result->mutable_view().begin<size_type>());
+    } else if (window_type == window_type::OPEN) {
+      thrust::copy_n(rmm::exec_policy(stream),
+                     cudf::detail::make_counting_transform_iterator(
+                       0,
+                       distance_kernel<T, OffsetType, op_t<Which, window_type::OPEN>>{
+                         d_length, d_offset, input_begin, input_end}),
+                     input.size(),
+                     result->mutable_view().begin<size_type>());
+    } else if (window_type == window_type::RIGHT_CLOSED) {
+      thrust::copy_n(rmm::exec_policy(stream),
+                     cudf::detail::make_counting_transform_iterator(
+                       0,
+                       distance_kernel<T, OffsetType, op_t<Which, window_type::RIGHT_CLOSED>>{
+                         d_length, d_offset, input_begin, input_end}),
+                     input.size(),
+                     result->mutable_view().begin<size_type>());
+    } else if (window_type == window_type::CLOSED) {
+      thrust::copy_n(rmm::exec_policy(stream),
+                     cudf::detail::make_counting_transform_iterator(
+                       0,
+                       distance_kernel<T, OffsetType, op_t<Which, window_type::CLOSED>>{
+                         d_length, d_offset, input_begin, input_end}),
+                     input.size(),
+                     result->mutable_view().begin<size_type>());
+    } else {
+      CUDF_FAIL("Unhandled window type.");
+    }
+    return result;
+  }
+
+  template <typename T, CUDF_ENABLE_IF(cudf::is_timestamp<T>())>
+  [[nodiscard]] std::unique_ptr<column> operator()(column_view const& input,
+                                                   scalar const& length,
+                                                   scalar const& offset,
+                                                   window_type const window_type,
+                                                   rmm::cuda_stream_view stream,
+                                                   rmm::device_async_resource_ref mr) const
+  {
+    using OffsetType = typename T::duration;
+    using ScalarType = duration_scalar<OffsetType>;
+    CUDF_EXPECTS(cudf::is_duration(length.type()), "Length and offset must be duration types.");
+    CUDF_EXPECTS(length.type().id() == type_to_id<OffsetType>(),
+                 "Length must have same the resolution as the input.");
```

Comment:
This is a tighter requirement than strictly necessary. We just need it to be the case that we can add `OffsetType` to `T` (so I think it should be OK for example having a `TIMESTAMP_MICROSECONDS` column and adding a `DURATION_SECONDS` offset/length).

But to fully handle that I think one needs triple type-dispatch over the three-tuple of `(input_type, length, offset)`, so I decided not to do that unless it turns out to be absolutely necessary.